src/objective/rank_xendcg_objective.hpp

```diff
@@ -0,0 +1,145 @@
+ */
+#ifndef LIGHTGBM_OBJECTIVE_RANK_XENDCG_OBJECTIVE_HPP_
+#define LIGHTGBM_OBJECTIVE_RANK_XENDCG_OBJECTIVE_HPP_
+
+#include <LightGBM/objective_function.h>
+#include <LightGBM/utils/common.h>
+#include <LightGBM/utils/random.h>
+
+#include <string>
+#include <vector>
+
+namespace LightGBM {
+/*!
+* \brief Implementation of the learning-to-rank objective function, XE_NDCG [arxiv.org/abs/1911.09798].
+*/
+class RankXENDCG: public ObjectiveFunction {
+ public:
+  explicit RankXENDCG(const Config& config) {
+    if (config.seed != 0) {
+      rand_ = new Random(config.seed);
+    } else {
+      rand_ = new Random();
+    }
+  }
+
+  explicit RankXENDCG(const std::vector<std::string>&) {
+    rand_ = new Random();
+  }
+
+  ~RankXENDCG() {
+  }
+  void Init(const Metadata& metadata, data_size_t) override {
+    // get label
+    label_ = metadata.label();
+    // get boundries
+    query_boundaries_ = metadata.query_boundaries();
+    if (query_boundaries_ == nullptr) {
+      Log::Fatal("RankXENDCG tasks require query information");
+    }
+    num_queries_ = metadata.num_queries();
+  }
+
+  void GetGradients(const double* score, score_t* gradients,
+                    score_t* hessians) const override {
+    #pragma omp parallel for schedule(guided)
+    for (data_size_t i = 0; i < num_queries_; ++i) {
+      GetGradientsForOneQuery(score, gradients, hessians, i);
+    }
+  }
+
+  inline void GetGradientsForOneQuery(
+      const double* score,
+      score_t* lambdas, score_t* hessians, data_size_t query_id) const {
+    // get doc boundary for current query
+    const data_size_t start = query_boundaries_[query_id];
+    const data_size_t cnt =
+      query_boundaries_[query_id + 1] - query_boundaries_[query_id];
+    // add pointers with offset
+    const label_t* label = label_ + start;
+    score += start;
+    lambdas += start;
+    hessians += start;
+
+    // Turn scores into a probability distribution using Softmax.
+    std::vector<double> rho(cnt);
+    Common::Softmax(score, &rho[0], cnt);
+
+    // Prepare a vector of gammas, a parameter of the loss.
+    std::vector<double> gammas(cnt);
+    for (data_size_t i = 0; i < cnt; ++i) {
+      gammas[i] = rand_->NextFloat();
+    }
+
+    // Skip query if sum of labels is 0.
+    float sum_labels = 0;
+    for (data_size_t i = 0; i < cnt; ++i) {
+      sum_labels += phi(label[i], gammas[i]);
+    }
+    if (sum_labels == 0) {
+      return;
+    }
+
+    // Approximate gradients and inverse Hessian.
+    // First order terms.
+    std::vector<double> L1s(cnt);
+    for (data_size_t i = 0; i < cnt; ++i) {
+      L1s[i] = -phi(label[i], gammas[i])/sum_labels + rho[i];
+    }
+    // Second-order terms.
+    std::vector<double> L2s(cnt);
+    for (data_size_t i = 0; i < cnt; ++i) {
+      for (data_size_t j = 0; j < cnt; ++j) {
+        if (i == j) continue;
+        L2s[i] += L1s[j] / (1 - rho[j]);
```

Comment:
Can't division by zero happen here (and in some other places)?

Reply:
A division by zero could only happen if \rho is 1. That could only happen if there is only a single document in the list (i.e., cnt is 1) -- that is because \rho is a probability distribution generated by Softmax. However, if cnt is 1, line 96 would cause the loop to exit.

So neither line 97 nor line 105 below should result in a division by zero. Does that make sense?

Reply:
> Does that make sense?

Absolutely! Thank you!